memory.go

```diff
@@ -0,0 +1,227 @@
+	if ro {
+		flags = unix.PROT_READ
+	}
+
+	// Map the bpf map memory over the Go heap. This will result in the following
+	// mmap layout in the process' address space (0xc000000000 is a span of Go
+	// heap), visualized using pmap:
+	//
+	// Address           Kbytes     RSS   Dirty Mode  Mapping
+	// 000000c000000000    1824     864     864 rw--- [ anon ]
+	// 000000c0001c8000       4       4       4 rw-s- [ anon ]
+	// 000000c0001c9000    2268      16      16 rw--- [ anon ]
+	//
+	// This will break up the Go heap, but as long as the runtime doesn't try to
+	// move our allocation around, this is safe for as long as we hold a reference
+	// to our allocated object.
+	//
+	// Use MAP_SHARED to make sure the kernel sees any writes we do, and MAP_FIXED
+	// to ensure the mapping starts exactly at the address we requested. If alloc
+	// isn't page-aligned, the mapping operation will fail.
+	if _, err = unix.MmapPtr(fd, 0, alloc, uintptr(size),
+		flags, unix.MAP_SHARED|unix.MAP_FIXED); err != nil {
+		return nil, fmt.Errorf("setting up memory-mapped region: %w", err)
+	}
+
+	// Set a finalizer on the heap allocation to undo the mapping before the span
+	// is collected and reused by the runtime. This has a few reasons:
+	//
+	//  - Avoid leaking memory/mappings.
+	//  - Future writes to this memory should never clobber a bpf map's contents.
+	//  - Some bpf maps are mapped read-only, causing a segfault if the runtime
+	//    reallocates and zeroes the span later.
+	runtime.SetFinalizer((*byte)(alloc), unmap(size))
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+	}
+
+	return mm, nil
+}
+
+// allocate returns an unsafe.Pointer to a page-aligned section of memory on the
+// Go heap, managed by the runtime. The given size is rounded up to the nearest
+// multiple of the system's page size to ensure we're given an allocation that
+// starts on a page boundary. The size of the resulting allocation is returned.
+func allocate(size int) (unsafe.Pointer, int, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, 0, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	// Request at least a full page from the runtime, otherwise the allocated span
+	// is likely not page-aligned, risking mapping over objects on another page.
+	// Since we use MAP_FIXED, the starting address of the mapping must be
+	// page-aligned on most architectures anyway.
+	size = internal.Align(size, os.Getpagesize())
+
+	// Get the address of the backing array of the slice and check if it's
+	// page-aligned to make sure it works with MAP_FIXED.
+	ptr := unsafe.Pointer(unsafe.SliceData(make([]byte, size)))
+	addr := int(uintptr(ptr))
+	if internal.Align(addr, os.Getpagesize()) != addr {
+		return nil, 0, fmt.Errorf("allocated memory is not page-aligned: %d", addr)
+	}
```

Comment:
It is surprising that we *reliably* get a page-aligned piece of memory here. Runtime manages various size classes differently, and thresholds are [subject to change](https://go.dev/doc/go1.22#runtime).

Should we get an unaligned chunk, we could retry with +1 page size and re-slice.

Reply:
Here's the current table of size classes: https://github.com/golang/go/blob/master/src/runtime/sizeclasses.go. I can't find this mentioned anywhere explicitly, but isn't this just an extension of typical struct alignment rules? If this slice would be a 4k struct instead, it would need to be 4k-aligned to make sure any pointers within the struct would be correctly 8-byte aligned by extension. I don't think an allocator is allowed to return allocations at unaligned offsets, but I could be wrong of course.

We could test this further by allocating some ballast and/or fragmenting the heap before running the tests a large number of times.

Reply:
>this slice would be a 4k struct instead, it would need to be 4k-aligned to make sure any pointers within the struct would be correctly 8-byte aligned by extension.

I don’t get why it needs to be 4k aligned. 8 byte alignment should be sufficient.

Reply:
Never mind my contrived example. :) If you request an 8-byte allocation, it needs to be aligned to 8 bytes. If you request a 4k allocation, it needs to be aligned to 4k bytes, etc.

Reply:
I asked on golang-nuts here: https://groups.google.com/g/golang-nuts/c/LzUMguyouXM/m/Gcngw8OuAAAJ. Will update here if any responses come in.

Reply:
tl;dr outcome of the thread:
- as expected, "why would anyone do this?"
- don't depend on any allocs being page-aligned, request an extra page to make sure we can align

The latter makes a lot of sense and is an easy change.